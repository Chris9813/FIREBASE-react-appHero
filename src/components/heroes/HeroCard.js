import React from 'react'
import { Link } from 'react-router-dom'
import { heroImages } from '../../helpers/heroImages'



export const HeroCard = ({
    id,
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
}) => {
    return (
        <div className = "row row-cols-2" style={{maxWidth: 540}}>
        <div className="card-body">
            <img src = {heroImages(`./${id}.jpg`).default} className="card-img-top" alt="{superhero}"/>
        </div>
        
            <div className ="card-body">
                <h5 className = "card-title">{superhero}</h5>
                <p className ="card-text-center">{alter_ego}</p>
            
            {
                (alter_ego !== characters)
                    && <p className ="card-text">{characters}</p>
            }
            <p className ="card-text">
                <small>{first_appearance}</small> 
            </p>
            <Link className = "link-info" to={`./hero/${id}`}>
                Mas...
            </Link>
            </div>
        </div>
        
    )
}