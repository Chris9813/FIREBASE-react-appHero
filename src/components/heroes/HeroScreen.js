import React, { useMemo } from 'react'
import { useParams, Redirect } from 'react-router-dom'
import {getHeroesById}  from '../selectors/getHeroesById'
//import batman from "../../assets/heroes/dc-batman.jpg"
const heroImages = require.context("../../assets/heroes", true)



export const HeroScreen = ({history}) => {
    const {heroeId} = useParams()
    const hero = useMemo(() => getHeroesById(heroeId), [heroeId]);

    if(!hero[0]){
        return <Redirect to ="/" />
    }

    const handleReturn = () => {
        if(history.length <= 2){
            history.push("/")
        } else{
            history.goBack();
        }
    }
    const heroes = hero[0]
    const {
    superhero,
    publisher,
    alter_ego,
    first_appearance,
    characters,
    } = heroes


    return (

    <div>
        <div className="row mt-5">
            <div className ="col-4">
                <img 
                    //src={`../assets/heroes/${heroeId}.jpg`}// desde public/assets
                    //src={batman}// con imp

                    src = {heroImages(`./${heroeId}.jpg`).default}
                    alt = {superhero}
                    className = "img-thumbnail animate__animated animate__backInRight"
                    />
            </div>
            <div className ="col-8">
                <h3>{superhero}</h3>
                <ul className = "list-group list-grup-flush">
                    <li className="list-group-item"> <b>Alter ego: </b>{alter_ego} </li>
                    <li className="list-group-item"> <b>Publisher: </b>{publisher} </li>
                    <li className="list-group-item"> <b>First appearence: </b>{first_appearance} </li>
                </ul>
                <h5>Characters</h5>
                <p>{characters}</p>
                <button className ="btn btn-outline-info"
                onClick = {handleReturn}
                >
                    Return
                </button>
            </div>
        </div>
    </div>
    )
}
