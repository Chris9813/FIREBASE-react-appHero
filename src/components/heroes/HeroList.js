import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard';


export const HeroList = ({publisher}) => {

    const heroes = useMemo(() => 
        getHeroesByPublisher(publisher), [publisher])

    return (
        <div className = "container">
            <div 
            className = "row row-cols-3 animate__animated animate__backInDown">
        {
            heroes.map(hero => (
                
            <div className = "col">
                <HeroCard 
                key = {hero.id}
                {...hero}
                />
            </div>
            ))
        }
        </div>
        </div>
    )
}
