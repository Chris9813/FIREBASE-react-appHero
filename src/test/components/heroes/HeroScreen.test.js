import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { HeroScreen } from "../../../components/heroes/HeroScreen"


describe('Pruebas en <HeroScreen />', () => {

    const historyMock = {
        length: 10,
        goBack: jest.fn(),
        push: jest.fn(),
        hero: ["asdasd"]
    }
    
    
    
    test('debe mostrar el componente correcctamente redirect si no hay argumentos en el url', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries = {["/hersdf"]}>
            <Route path="/hero/:heroeId" component = {HeroScreen}/>
            </MemoryRouter>)
        expect(wrapper.text()).toBe("");
        expect(wrapper).toMatchSnapshot()
        //expect(wrapper.find("Redirect").exists()).toBe(true)
    })
    
    test('debe mostrar un hero si el parametro existe ', () => {
    const wrapper = mount(
        <MemoryRouter initialEntries = {["/hero/marvel-captain"]}>
        <Route path="/hero/:heroeId" component = {HeroScreen}/>
        </MemoryRouter>)
        expect(wrapper.find(".row").exists()).toBe(true)
    })
    
    test('debe regrear a la pantalla anterior con push ', () => {
        
        const historyMock = {
            length: 1,
            goBack: jest.fn(),
            push: jest.fn(),
        }
        const wrapper = mount(
            <MemoryRouter initialEntries = {["/hero/marvel-captain"]}>
            <Route path="/hero/:heroeId" 
            component = {(props) =>
            <HeroScreen history = {historyMock}/>}/>
            </MemoryRouter>
        )
        const button = wrapper.find("button");
        button.simulate("click");
        expect(historyMock.push).toHaveBeenCalledWith("/");
        expect(historyMock.goBack).not.toHaveBeenCalled();
    })

    test('debe regresar a la ventana anterior GOBACK', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries = {["/hero/marvel-captain"]}>
            <Route path="/hero/:heroeId" 
            component = {(props) =>
            <HeroScreen history = {historyMock}/>}/>
            </MemoryRouter>
        )
        const button = wrapper.find("button");
        button.simulate("click");
        expect(historyMock.push).not.toHaveBeenCalled();
        expect(historyMock.goBack).toHaveBeenCalled();
    })
    
    test('debe de llamar el redirect si el hero no exisste ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries = {["/hero/12345678"]}>
            <Route path="/hero/:heroeId" 
            component = {(props) =>
            <HeroScreen history = {historyMock}/>}/>
            </MemoryRouter>
        )
        expect(wrapper.text()).toBe("")
    })
    
    
})
