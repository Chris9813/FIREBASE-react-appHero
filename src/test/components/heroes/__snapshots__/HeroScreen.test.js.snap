// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen /> debe mostrar el componente correcctamente redirect si no hay argumentos en el url 1`] = `""`;
